cmake/patch_llvm.py

```diff
@@ -0,0 +1,132 @@
+    patch_names.sort()
+
+    print(f"Found {len(patch_names)} patches to apply:")
+    for patch_name in patch_names:
+        print(patch_name)
+
+    if args.method == "am":
+        merge_args = ["git", "am", "-k", "--ignore-whitespace", "--3way"]
+        for patch_name in patch_names:
+            merge_args.append(os.path.join(args.patchdir, patch_name))
+        p = subprocess.run(
+            merge_args, cwd=args.llvm_dir, capture_output=True, text=True
+        )
+        print(p.stdout)
+        print(p.stderr)
+
+        if p.returncode == 0:
+            print(f"All patches applied.")
+            sys.exit(0)
+        if args.restore_on_fail:
+            # Check that the operation can be aborted.
+            if (
+                'To restore the original branch and stop patching, run "git am --abort".'
+                in p.stdout
+            ):
+                print("Aborting git am...")
+                subprocess.run(["git", "am", "--abort"], cwd=args.llvm_dir, check=True)
+                sys.exit(2)
+        sys.exit(1)
+    else:
+        applied_patches = []
+        for patch_name in patch_names:
+            patch_file = os.path.join(args.patchdir, patch_name)
+            print(f"Checking {patch_file}...")
+            # Check that the patch applies before trying to apply it.
+            apply_check_args = [
+                "git",
+                "apply",
+                "--ignore-whitespace",
+                "--3way",
+                "--check",
+                patch_file,
+            ]
+            p_check = subprocess.run(apply_check_args, cwd=args.llvm_dir)
+
+            if p_check.returncode == 0:
+                # Patch will apply.
+                print(f"Applying {patch_file}...")
+                apply_args = [
+                    "git",
+                    "apply",
+                    "--ignore-whitespace",
+                    "--3way",
+                    patch_file,
+                ]
+                apply_args = subprocess.run(apply_args, cwd=args.llvm_dir, check=True)
+                applied_patches.append(patch_file)
+            else:
+                # Patch won't apply.
+                print(f"Unable to apply {patch_file}")
+                if args.restore_on_fail:
+                    # Remove any patches that have already been applied.
+                    while len(applied_patches) > 0:
+                        r_patch = applied_patches.pop()
+                        print(f"Reversing {r_patch}...")
+                        reverse_args = [
+                            "git",
+                            "apply",
+                            "--ignore-whitespace",
+                            "--3way",
+                            "--reverse",
+                            r_patch,
+                        ]
+                        p_check = subprocess.run(
+                            reverse_args, cwd=args.llvm_dir, check=True
+                        )
```

Comment:
Would it not be simpler to just remember the commit before we started applying patches, and do a git reset to that if a patch fails? One liner, and no need to keep track of applied patches.

Reply:
A reset would be simple, but would also reset all changes, including any other patch sets that were previously applied or any local changes by the user. Rewinding is more involved, but gets you back to the same state as you started with.